VideoLocker/src/main/java/org/edx/mobile/view/CourseBaseActivity.java

```diff
@@ -119,22 +124,57 @@ protected void onPause() {
     @Override
     protected void onDestroy() {
         super.onDestroy();
+        if (getHierarchyTask != null) {
+            getHierarchyTask.cancel(true);
+            getHierarchyTask = null;
+        }
     }
 
 
     @Override
     public void onSaveInstanceState(Bundle outState) {
-        if ( courseData != null)
-            outState.putSerializable(Router.EXTRA_ENROLLMENT, courseData);
-        if ( courseComponentId != null )
-            outState.putString(Router.EXTRA_COURSE_COMPONENT_ID, courseComponentId);
         super.onSaveInstanceState(outState);
+        outState.putSerializable(Router.EXTRA_ENROLLMENT, courseData);
+        outState.putString(Router.EXTRA_COURSE_COMPONENT_ID, courseComponentId);
     }
 
     protected void restore(Bundle savedInstanceState) {
-        if (savedInstanceState != null) {
-            courseData = (EnrolledCoursesResponse) savedInstanceState.getSerializable(Router.EXTRA_ENROLLMENT);
-            courseComponentId = savedInstanceState.getString(Router.EXTRA_COURSE_COMPONENT_ID);
+        courseData = (EnrolledCoursesResponse) savedInstanceState.getSerializable(Router.EXTRA_ENROLLMENT);
+        courseComponentId = savedInstanceState.getString(Router.EXTRA_COURSE_COMPONENT_ID);
+
+        if (courseComponentId == null) {
+            getHierarchyTask = new GetCourseStructureTask(this, courseData.getCourse().getId()) {
+                @Override
+                public void onSuccess(CourseComponent courseComponent) {
+                    if (courseComponent != null) {
+                        courseComponentId = courseComponent.getId();
+                        // The getHierarchyTask variable is set to null in onDestroy(), so this,
+                        // check ensures that the Activity has not been destroyed yet.
+                        if (getHierarchyTask != null) {
```

Comment:
I actually wanted to check the state of the isCancelled() method, as the task being cancelled was a more generic indicator that the listener was not interested in the data anymore. However, it just delegates all the canceling methods to the Future instance that it uses, and Future canceling has no effect after it’s task has been completed (but potentially before we get the callback on the main thread). (AsyncTask is much better for this, as it does not lose this state.) As such setting the variable to null was used as a proxy to retain the canceled state without having to extend the class to provide a framework for it myself, although I guess for a more generic check I should check for "if (getHierarchyTask != this)” instead.

Still, in this case since we’re only canceling on the destruction of the Fragment, we could use a flag for that purpose as well.
> On 17-Sep-2015, at 1:57 am, Akiva Leffert <notifications@github.com> wrote:
> 
> In VideoLocker/src/main/java/org/edx/mobile/view/CourseBaseActivity.java <https://github.com/edx/edx-app-android/pull/360#discussion_r39684142>:
> 
> >      }
> >  
> >      protected void restore(Bundle savedInstanceState) {
> > -        if (savedInstanceState != null) {
> > -            courseData = (EnrolledCoursesResponse) savedInstanceState.getSerializable(Router.EXTRA_ENROLLMENT);
> > -            courseComponentId = savedInstanceState.getString(Router.EXTRA_COURSE_COMPONENT_ID);
> > +        courseData = (EnrolledCoursesResponse) savedInstanceState.getSerializable(Router.EXTRA_ENROLLMENT);
> > +        courseComponentId = savedInstanceState.getString(Router.EXTRA_COURSE_COMPONENT_ID);
> > +
> > +        if (courseComponentId == null) {
> > +            getHierarchyTask = new GetCourseStructureTask(this, courseData.getCourse().getId()) {
> > +                @Override
> > +                public void onSuccess(CourseComponent courseComponent) {
> > +                    if (courseComponent != null) {
> > +                        courseComponentId = courseComponent.getId();
> > +                        // The getHierarchyTask variable is set to null in onDestroy(), so this,
> > +                        // check ensures that the Activity has not been destroyed yet.
> > +                        if (getHierarchyTask != null) {
> Is it safer to just have a destroyed flag? That way if the task logic gets refactored at all, no one has to ensure the contract from this comment is still valid? Or is this too linked to that logic anyway?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/edx/edx-app-android/pull/360/files#r39684142>.
>